Modifying some nodes with the lexicalPreservation enabled throws an exception

I'm trying to reorder the parameters in a constructor creation expr.
The first thing I've tried is to reorder the NodeList => Not sucessful (Looked normal since the parent is not aware of the sorting)
Second thing I've tried it to replace it manually, see code below:
```java
      Map<Integer, Expression> positionToParameterMap = new HashMap<>();
      NodeList<Expression> arguments = constructorCallExpr.getArguments();
      for (int i = 0; i < arguments.size(); i++) {
          positionToParameterMap.put(i, arguments.get(i));
      }
      for (Pair<Integer, Integer> deltaPair : constructorChanges.deltaChanges) {
          Expression expression = positionToParameterMap.get(deltaPair.getValue()); // map to element at previous position
          constructorCallExpr.setArgument(deltaPair.getKey(), expression); // Replace element at new position with element from previous position
      }
```

This code above throws:

```
Caused by: java.lang.IllegalArgumentException: fromIndex(116) > toIndex(73)
        at java.util.SubList.<init>(AbstractList.java:624)
        at java.util.AbstractList.subList(AbstractList.java:484)
        at com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator$CalculatedSyntaxModel.sub(LexicalDifferenceCalculator.java:66)
        at com.github.javaparser.printer.lexicalpreservation.DifferenceElementCalculator.calculate(DifferenceElementCalculator.java:144)
        at com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator.calculateListReplacementDifference(LexicalDifferenceCalculator.java:132)
        at com.github.javaparser.printer.lexicalpreservation.LexicalPreservingPrinter$Observer.concreteListReplacement(LexicalPreservingPrinter.java:307)
        at com.github.javaparser.ast.observer.PropagatingAstObserver.listReplacement(PropagatingAstObserver.java:82)
        at com.github.javaparser.ast.NodeList.lambda$notifyElementReplaced$3(NodeList.java:474)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at com.github.javaparser.ast.NodeList.notifyElementReplaced(NodeList.java:474)
        at com.github.javaparser.ast.NodeList.set(NodeList.java:147)
        at com.github.javaparser.ast.nodeTypes.NodeWithArguments.setArgument(NodeWithArguments.java:55)
```

Any idea why? Am I doing something I shouldn't be doing? If no, is there a workaround for this?

--- a/javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/Issue2517Test.java
+++ b/javaparser-core-testing/src/test/java/com/github/javaparser/printer/lexicalpreservation/Issue2517Test.java
@@ -1,0 +1,70 @@
+package com.github.javaparser.printer.lexicalpreservation;
+
+import static com.github.javaparser.utils.TestUtils.assertEqualsStringIgnoringEol;
+
+import java.util.StringJoiner;
+
+/*
+ * Copyright (C) 2007-2010 Júlio Vilmar Gesser.
+ * Copyright (C) 2011, 2013-2019 The JavaParser Team.
+ *
+ * This file is part of JavaParser.
+ *
+ * JavaParser can be used either under the terms of
+ * a) the GNU Lesser General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ * b) the terms of the Apache License
+ *
+ * You should have received a copy of both licenses in LICENCE.LGPL and
+ * LICENCE.APACHE. Please refer to those files for details.
+ *
+ * JavaParser is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+import org.junit.jupiter.api.Test;
+
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.expr.Expression;
+import com.github.javaparser.ast.expr.ObjectCreationExpr;
+import com.github.javaparser.javadoc.Javadoc;
+import com.github.javaparser.javadoc.description.JavadocDescription;
+
+public class Issue2517Test extends AbstractLexicalPreservingTest {
+
+	@Test
+	public void test() {
+
+		considerCode("public class A {\n" +
+	            "    public A(String a , String b) {\n" +
+	            "    }\n" +
+	            "    public static A m() {\n" +
+	            "      return new A(\"a\",\"b\");\n" +
+	            "    }\n" +
+	            "}");
+		
+		String expected =
+				"public class A {\n" +
+			    "    public A(String a , String b) {\n" +
+			    "    }\n" +
+			    "    public static A m() {\n" +
+			    "      return new A(\"b\", \"a\");\n" +
+			    "    }\n" +
+			    "}";
+
+	    ObjectCreationExpr cd = cu.findFirst(ObjectCreationExpr.class).get();
+	    NodeList<Expression> args = cd.getArguments();
+	    Expression a1 = args.get(0);
+	    Expression a2 = args.get(1);
+	    NodeList<Expression> newArgs = new NodeList<>(a2, a1);
+	    cd.setArguments(newArgs);
+
+	    assertEqualsStringIgnoringEol(expected, LexicalPreservingPrinter.print(cu));
+	}
+}

--- a/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/DifferenceElementCalculator.java
+++ b/javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/DifferenceElementCalculator.java
@@ -24,9 +24,11 @@ import com.github.javaparser.ast.Node;
 import com.github.javaparser.ast.body.VariableDeclarator;
 import com.github.javaparser.ast.type.Type;
 import com.github.javaparser.printer.concretesyntaxmodel.*;
+import com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator.CalculatedSyntaxModel;
 import com.github.javaparser.printer.lexicalpreservation.LexicalDifferenceCalculator.CsmChild;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
 
@@ -171,21 +173,45 @@ class DifferenceElementCalculator {
         int originalIndex = 0;
         int afterIndex = 0;
         int commonChildrenIndex = 0;
+        // The algorithm is based on common child elements. 
+        // It first analyzes the elements preceding this child. 
+        // Then it keeps the common element and continues the analysis between the element 
+        // following this child and the common element in the list.
         while (commonChildrenIndex < commonChildren.size()) {
             ChildPositionInfo child = commonChildren.get(commonChildrenIndex++);
             // search the position of the node "child" in the original list of cms element
             int posOfNextChildInOriginal = childrenInOriginal.stream().filter(i -> i.equals(child)).map(i -> i.position).findFirst().get();
             // search the position of the node "child" in the modified list of cms element
             int posOfNextChildInAfter = childrenInAfter.stream().filter(i -> i.equals(child)).map(i -> i.position).findFirst().get();
+            // Imagine that the common elements has been moved, for example in the case where the parameters of a method are reversed
+			// In this case the afterIndex will be greater than the position of the child in
+			// the list
+			// For example : if {@code new Foo(a, b)} become {@code new Foo(b, a)}
+			// Nota: in this example there is 3 child elements Foo, 'a' and 'b', others are tokens
+			// In the orginal list the child element 'a' is at the position 5 and the
+			// element 'b' is at the position 8
+			// After reverting the list of parameters the child element 'a' is at the
+			// position 8 and the element 'b' is at the position 5
+			// When we deal with element 'b', it is in 5th position in the list after the
+			// modification but the previous position in the list was that of element 'a'.
             if (originalIndex < posOfNextChildInOriginal || afterIndex < posOfNextChildInAfter) {
-                elements.addAll(calculateImpl(original.sub(originalIndex, posOfNextChildInOriginal), after.sub(afterIndex, posOfNextChildInAfter)));
+            	// defines the sublist of elements located before the common element  
+            	CalculatedSyntaxModel originalSub = originalIndex < posOfNextChildInOriginal ? original.sub(originalIndex, posOfNextChildInOriginal) : new CalculatedSyntaxModel(Collections.EMPTY_LIST);
+            	CalculatedSyntaxModel afterSub = afterIndex < posOfNextChildInAfter ? after.sub(afterIndex, posOfNextChildInAfter) : new CalculatedSyntaxModel(Collections.EMPTY_LIST);
+                elements.addAll(calculateImpl(originalSub, afterSub));
             }
-            elements.add(new Kept(new CsmChild(child.node)));
-            originalIndex = posOfNextChildInOriginal + 1;
-            afterIndex = posOfNextChildInAfter + 1;
+            if (afterIndex <= posOfNextChildInAfter) {
+            	elements.add(new Kept(new CsmChild(child.node)));
+            } else {
+            	elements.add(new Removed(new CsmChild(child.node)));
+            }
+            originalIndex = originalIndex <= posOfNextChildInOriginal ? posOfNextChildInOriginal + 1 : originalIndex;
+            afterIndex = afterIndex <= posOfNextChildInAfter ? posOfNextChildInAfter + 1 : afterIndex;
         }
         if (originalIndex < original.elements.size() || afterIndex < after.elements.size()) {
-            elements.addAll(calculateImpl(original.sub(originalIndex, original.elements.size()), after.sub(afterIndex, after.elements.size())));
+        	CalculatedSyntaxModel originalSub = originalIndex < original.elements.size() ? original.sub(originalIndex, original.elements.size()) : new CalculatedSyntaxModel(Collections.EMPTY_LIST);
+        	CalculatedSyntaxModel afterSub = afterIndex < after.elements.size() ? after.sub(afterIndex, after.elements.size()) : new CalculatedSyntaxModel(Collections.EMPTY_LIST);
+            elements.addAll(calculateImpl(originalSub, afterSub));
         }
         return elements;
     }

